### PR TITLE
Videoprocessor updates

### PR DESCRIPTION
**Pull request by charliesantos**
**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

See commit history for details on the changes.

### CODE REVIEW

File: lib/media/track/localmediatrack.js [hunk 1]
```diff
@@ -189,12 +189,19 @@ function mixinLocalMediaTrack(AudioOrVideoTrack) {
...
+        this._unprocessedTrack = mediaStreamTrack;
+        return Promise.resolve();
+      })() :
+        this._trackSender.setMediaStreamTrack(mediaStreamTrack).catch(error => {
+          this._log.warn('setMediaStreamTrack failed:', { error, mediaStreamTrack });
+        }))
+        .then(() => {
+          this._initialize();
+          this._getAllAttachedElements().forEach(el => this._attach(el));
+        });
```
Comment (manjeshbhargav) [inline lib/media/track/localmediatrack.js:204]: This can be:
```js
return this._unprocessedTrack ? Promise.resolve().then(() => {
  this._unprocessedTrack = mediaStreamTrack;
}) : this._trackSender.setMediaStreamTrack(mediaStreamTrack).catch(error => {
  this._log.warn('setMediaStreamTrack failed:', { error, mediaStreamTrack });
})).then(() => {
  this._initialize();
  this._getAllAttachedElements().forEach(el => this._attach(el));
});
```

File: lib/media/track/localmediatrack.js [hunk 2]
```diff
@@ -189,12 +189,19 @@ function mixinLocalMediaTrack(AudioOrVideoTrack) {
...
       this._stop();
 
-      return this._trackSender.setMediaStreamTrack(mediaStreamTrack).catch(error => {
-        this._log.warn('setMediaStreamTrack failed:', { error, mediaStreamTrack });
-      }).then(() => {
-        this._initialize();
-        this._getAllAttachedElements().forEach(el => this._attach(el));
-      });
+      // If there's an unprocessedTrack, this means RTCRtpSender has
+      // the processedTrack already set, we don't want to replace that.
```
Comment (manjeshbhargav) [inline lib/media/track/localmediatrack.js:193]: Please start any explanatory comments with `// NOTE(csantos):`.

File: lib/media/track/localvideotrack.js [hunk 3]
```diff
@@ -210,9 +210,7 @@ class LocalVideoTrack extends LocalMediaVideoTrack {
    * the <code>mediaStreamTrack</code> property. If you want to listen to events on
    * the MediaStreamTrack directly, please do so in the "started" event handler. Also,
    * the {@link LocalVideoTrack}'s ID is no longer guaranteed to be the same as the
-   * underlying MediaStreamTrack's ID. Additionally, the {@link LocalVideoTrack}'s `processor`
-   * and `processedTrack` properties will be null while the restart is in progress and will
-   * become available again once restart is complete.
```
Comment (manjeshbhargav) [inline lib/media/track/localvideotrack.js:215]: Why is this removed?

File: lib/media/track/mediatrack.js [hunk 4]
```diff
@@ -131,7 +131,8 @@ class MediaTrack extends Track {
     if (this._dummyEl) {
       this._dummyEl.muted = true;
       this._dummyEl.oncanplay = this._start.bind(this, this._dummyEl);
-      this._attach(this._dummyEl);
+      // We always want to attach the original mediaStreamTrack for dummyEl
```
Comment (manjeshbhargav) [inline lib/media/track/mediatrack.js:134]: Please start any explanatory comments with `// NOTE(csantos):`.

File: lib/media/track/videotrack.js [hunk 5]
```diff
@@ -224,6 +239,20 @@ class VideoTrack extends MediaTrack {
     }
     this._log.debug('Adding VideoProcessor to the VideoTrack', processor);
 
+    if (!this._unmuteHandler) {
+      this._unmuteHandler = () => {
+        this._log.debug('mediaStreamTrack unmuted');
+        // On certain scenarios where mediaStreamTrack is coming from muted to unmuted state,
```
Comment (manjeshbhargav) [inline lib/media/track/videotrack.js:251]: Please start any explanatory comments with `// NOTE(csantos):`.

File: lib/media/track/videotrack.js [hunk 6]
```diff
@@ -110,12 +110,12 @@ class VideoTrack extends MediaTrack {
 
     let startTime = Date.now();
     let processFramePeriodMs;
-    const { frameRate } = this.mediaStreamTrack.getSettings();
-    const capturePeriodMs = Math.floor(1000 / frameRate);
 
     this._dummyEl.play().then(() => {
       const captureFrame = cb => {
         clearTimeout(this._captureTimeoutId);
+        const { frameRate = 24 } = this.mediaStreamTrack.getSettings();
```
Comment (manjeshbhargav) [inline lib/media/track/videotrack.js:117]: Can you move this to `lib/util/constants.js`?

**Review by manjeshbhargav — APPROVED** [review]

File: lib/media/track/localvideotrack.js (hunk 3 above)
Comment (charliesantos) [reply lib/media/track/localvideotrack.js:215]: ([VIDEO-4028](https://issues.corp.twilio.com/browse/VIDEO-4028)) The initial implementation of track.restart with VideoProcessor is that, VideoTrack.processor and VideoTrack.processedTrack are null after calling track.restart. This is because track.restart is asynchronous and we were waiting for it to finish before re-adding the processors back.

With this change ([VIDEO-4028](https://issues.corp.twilio.com/browse/VIDEO-4028), see also commit), processor and processedTrack properties now not null during track.restart.

File: lib/media/track/mediatrack.js [hunk 7]
```diff
@@ -131,7 +131,8 @@ class MediaTrack extends Track {
     if (this._dummyEl) {
       this._dummyEl.muted = true;
       this._dummyEl.oncanplay = this._start.bind(this, this._dummyEl);
-      this._attach(this._dummyEl);
+      // NOTE(csantos): We always want to attach the original mediaStreamTrack for dummyEl
+      this._attach(this._dummyEl, this.mediaStreamTrack);
```
Comment (manjeshbhargav) [inline lib/media/track/mediatrack.js:135]:
```suggestion

      // NOTE(csantos): We always want to attach the original mediaStreamTrack for dummyEl
      this._attach(this._dummyEl, this.mediaStreamTrack);

```

File: lib/media/track/videotrack.js [hunk 8]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 
 const MediaTrack = require('./mediatrack');
+const DEFAULT_FRAME_RATE = require('../../util/constants').DEFAULT_FRAME_RATE;
```
Comment (manjeshbhargav) [inline lib/media/track/videotrack.js:4]:
```suggestion
const { DEFAULT_FRAME_RATE } = require('../../util/constants');
```

**Review by manjeshbhargav — APPROVED** [review]
